File: 1/2020.09.15/239.Sliding Window Maximum/java/solution.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
		if (nums == null || k <= 0) {
			return new int[0];
		}
		int n = nums.length;
		int[] res = new int[n-k+1];
		Deque<Integer> q = new LinkedList<>();
		for (int i = 0; i < n; i++) {
			if (!q.isEmpty() && q.peek() < i-k+1) {
				q.poll();
			}
			while (!q.isEmpty() && nums[q.peekLast()] <= nums[i]) {
				q.pollLast();
			}
			q.offer(i);
			if (i >= k - 1) {
				res[i-k+1] = nums[q.peek()];
			}
		}
		return res;
    }

    public int[] maxSlidingWindow(int[] nums, int k) {
        // dp
        if (nums == null || k == 0) {
            return new int[0];
        }
        int n = nums.length;
        int[] leftMax = new int[n];
        int[] rightMax = new int[n];
        leftMax[0] = nums[0];
        rightMax[n-1] = nums[n-1];
        for (int i = 1; i < n; i++) {
            leftMax[i] = i % k == 0 ? nums[i] : Math.max(leftMax[i-1], nums[i]);
            int j = n - i - 1;
            rightMax[j] = (j+1) % k == 0 ? nums[j] : Math.max(rightMax[j+1], nums[j]);
        }
        int[] res = new int[n-k+1];
        for (int i = 0; i < res.length; i++) {
            res[i] = Math.max(rightMax[i], leftMax[i+k-1]);
        }
        return res;
    }
}

File: 1/2020.09.15/239.Sliding Window Maximum/python/solution.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        q = []
        res = []
        n = len(nums)
        for i in range(n):
            if q != [] and q[0] < i - k + 1:
                q.pop(0)
            while q != [] and nums[q[-1]] <= nums[i]:
                q.pop()
            q.append(i)
            if i >= k - 1:
                res.append(nums[q[0]])
        return res

    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        # dp
        n = len(nums)
        left = [None for _ in range(n)]
        right = [None for _ in range(n)]
        left[0] = nums[0]
        right[-1] = nums[-1]
        for i in range(1, n):
            left[i] = nums[i] if i % k == 0 else max(left[i-1], nums[i])
            j = n - i - 1
            right[j] = nums[j] if (j+1) % k == 0 else max(right[j+1], nums[j])
        res = []
        for i in range(n-k+1):
            res.append(max(right[i], left[i+k-1]))
        return res
